les/pruner.go

```diff
@@ -0,0 +1,114 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package les
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// pruner is responsible for pruning historical light chain data.
+type pruner struct {
+	indexers []*core.ChainIndexer
+	closeCh  chan struct{}
+	wg       sync.WaitGroup
+}
+
+// newPruner returns a light chain pruner instance.
+func newPruner(indexers ...*core.ChainIndexer) *pruner {
+	pruner := &pruner{
+		indexers: indexers,
+		closeCh:  make(chan struct{}),
+	}
+	pruner.wg.Add(1)
+	go pruner.loop()
+	return pruner
+}
+
+// close notifies all background goroutines belonging to pruner to exit.
+func (p *pruner) close() {
+	close(p.closeCh)
+	p.wg.Wait()
+}
+
+// loop periodically queries the status of chain indexers and prunes useless
+// historical chain data. Notably, whenever Geth restarts, it will iterate
+// all historical sections even they don't exist at all(below checkpoint) so
+// that light client can prune cached chain data that was ODRed after pruning
+// that section.
+func (p *pruner) loop() {
+	defer p.wg.Done()
+
+	var (
+		last = ^uint64(0) // -1
+
+		// checkTicker is the ticker used to check whether there is a new section
+		// for pruning.
+		checkTicker = time.NewTicker(5 * time.Second)
+
+		// cleanTicker is the ticker used to trigger a history clean 2 times a day.
+		cleanTicker = time.NewTicker(12 * time.Hour)
+	)
+	// pruning finds the sections that have been processed by all indexers
+	// and deletes all historical chain data.
+	// Note, if some indexers don't support pruning(e.g. eth.BloomIndexer),
+	// pruning operations can be silently ignored.
+	pruning := func(cleanHistory bool) {
+		min := uint64(math.MaxUint64)
+		for _, indexer := range p.indexers {
+			sections, _, _ := indexer.Sections()
+			if sections < min {
+				min = sections
+			}
+		}
+		// Always keep the latest section data in database.
+		if min < 2 || len(p.indexers) == 0 {
+			return
+		}
+		if cleanHistory {
+			for _, indexer := range p.indexers {
+				indexer.PruneSections(0, min-2)
+			}
+			return
+		}
+		for i := last + 1; i < min-1; i++ {
+			for _, indexer := range p.indexers {
+				if err := indexer.PruneSection(i); err != nil {
+					log.Debug("Prune history data failed", "section", i, "err", err)
+					return
+				}
+			}
+			last = i
+		}
+	}
+	// Initial a first pruning task when setup which can clean the whole database.
+	pruning(false)
```

Comment:
Since this will do a "full" pruning at every startup anyway I think you could simplify this logic, drop the differential pruning option and just do a full pruning at startup and then every 12 hours. Sections are processed every 5-6 days, it does not matter really if you do the next pruning in a few seconds or in the next 12 hours. Then you could also drop PruneSection. PruneSections could have a single parameter and always start from block 1 (which is not more expensive than pruning a single section because you are using an iterator).

Reply:
Fixed.